README.md
Introduktion C++ Laboration 5
=============================

Uppgift 1a: Sök och sortera en array

Uppgift 1b: Sök och sortera en vector

Uppgift 2: Binärsökning i en array


Changelog.txt
Ja se på tusan. När jag kompilerade med GCC och la till -W -Wall -Wextra -pedantic -std=c++11
så dyker kompileringsfelet upp. Variabla arraylängder är tydligen en extension till GCC som
ska dyka upp i det nya C++14. Nu har jag byggt mitt program med hänsyn till GCC så den snabba
lösningen är att helt enkelt deklarera en konstant array och ta bort användarinmatningen. Det
var ju inget krav för den här uppgiften att användaren skulle bestämma storleken utan endast
ett eget tillägg. Nu bör koden kompilera.

Något fel om time och time.h eller ctime lyckas jag dock inte få fram oavsett flaggor.

lab5_upg2.cpp
// lab5_upg2.cpp
// Program för att utföra binärsökning på en array
// Thomas Nordenmark 2013-12-12
// Ver 0.1
//------------------------------------------------------------------------------
#include <random> // default_random_engine
#include <functional> // bind
#include <iostream> // cin, cout
#include <ctime> // time
using namespace std;

// ======================
// Funktionsdeklarationer
// ======================
const int MAX = 10;

// ======================
// Funktionsdeklarationer
// ======================
void bubbleSort(int arr[], int MAX); // Bubble sort
int binSearch(int arr[], int key, int MAX);

// ============
// Huvudprogram
// ============
int main()
{
    //int max; // Arraystorlek
    int range; // Maxtalet för slumpgeneratorn

    // Fråga efter och läs in maxtalet för slumpgeneratorn
    cout << "Mata in maxtalet för slumpgeneratorn: ";
    cin >> range;

    // Initiera slumpdistributörsobjektet generator
    default_random_engine generator(static_cast<unsigned>(time(0)));
    uniform_int_distribution<int> random(1, range); // Slumpa i itervallet 1-6
    // Slå ihop slumpgeneratorn till en funktion för enklare användning
    auto slumpaTal = bind(random, generator);

    int arr[MAX]; // Arrayen
    int i; // Iterator för loopar
    int key; // Det eftersökta talet
    int index; // Det eftersökta talets plats i arrayen

    cout << endl << "Arrayen osorterad:" << endl;
    for(i = 0; i < MAX; i++)
    {
        arr[i] = slumpaTal();
        cout << "a[" << i <<"] = " << arr[i] << endl;
    }


    // Sortera arrayen, en förutsättning för binärsökning
    bubbleSort(arr, MAX);


    // Skriv ut arrayen efter sortering
    cout << endl << "Arrayen sorterad i stigande ordning:" << endl;

    for(i = 0; i < MAX; i++)
    {
        cout << "a[" << i <<"] = " << arr[i] << endl;
    }


    // Fråga efter och läs in nummer att söka efter
    cout << endl << "Mata in ett nummer att söka efter mellan 1 och " << range << ": ";
    cin >> key;

    // Anropa binärsökningsfunktionen med array, sökord och storlek som parametrar
    index = binSearch(arr, key, MAX);


    // Presentera resultatet av binärsökningen
    if(index != -1)
        cout << "Det eftersökta numret finns på index: " << index << endl;
    else
        cout << "Det eftersökta numret hittades inte." << endl;


    return 0;
}


// =====================
// Funktionsdefinitioner
// =====================

// Bubble sort
void bubbleSort(int arr[], int max)
{
    int i, j; // Iteratorer för looparna
    int temp; // Temporär lagringsplats för sortering

    // Yttre loop
    for(i = 0; i < max; i++)
    {
        // Inre loop
        for(j = 0; j < max - 1; j++)
        {
            if(arr[j] > arr[j + 1])
            {
                temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

// Binär sökning
int binSearch(int arr[], int key, int max)
{
    int low = 0;
    int high = max - 1;
    int middle;

    while(low <= high)
    {
        middle = low + (high - low)/2; // Hitta arrayens mittenvärde

        // Om det eftersöka numret är högre än mittenvärdet, sök i den högre halvan
        if(key > arr[middle])
            low = middle + 1;
        // Om det eftersökta numret är lägre än mittenvärdet, sök i lägre halvan
        else if(key < arr[middle])
            high = middle - 1;
        // Om det eftersökta talet är mittenvärdet, returnera det
        else
            return middle;
    }

    return -1;
}


lab5_upg1_vector/lab5_upg1_vector.cpp
// lab5_upg1_vector.cpp
// Program för att utföra ett antal operationer på en vector
// Thomas Nordenmark 2013-12-012
// Ver 0.1
//------------------------------------------------------------------------------
#include <iostream> // cin, cout
#include <random> // default_random_engine
#include <iomanip> // setw, setprecision
#include <vector> // vector
#include <algorithm> // generate, sort, count etc
#include <functional> // bind
using namespace std;

// =========================
// Deklaration av konstanter
// =========================
const int MAX=600; // Max platser i vectorn
const int FREK=100; // Antal platser i frekvensvector samt slumpgeneratorns max

// ============
// Huvudprogram
// ============
int main()
{
    // Initiera slumpdistributörsobjektet generator
    default_random_engine generator(static_cast<unsigned>(time(0)));
    uniform_int_distribution<int> random(1, FREK); // Slumpa i itervallet 1-100
    // Slå ihop slumpgeneratorn till en funktion för enklare användning
    auto slumpaTal = bind(random, generator);

    // Iteratorer för loopar
    int i;
    //int j; // Behövs för utskriftsmetod 1

    // Fyll ivec1 med 600 slumpade tal mellan 1 och 100
    cout << "Slumpar 600 tal mellan 1 och 100 till ivec1." << endl << endl;
    vector<int> ivec1(MAX, 0); // Vector nr 1
    generate(ivec1.begin(), ivec1.end(), slumpaTal);


    // Kopiera ivec1 till ivec2
    vector<int> ivec2; // Kopia av ivec1
    ivec2 = ivec1;


    // Beräkna medelvärdet av alla tal i ivec2
    int sum = 0;
    for(auto i: ivec2)
        sum += i;

    double medel = 0;
    medel = (double)sum / (double)MAX; // Typecasta till double för högre precision
    cout << "Vectorns medelvärde är: " << fixed << setprecision(2) << medel << endl << endl; // Skriv ut medel med 2 decimaler


    // Sök efter minsta och största talet i ivec2
    int smallest = *(min_element(ivec2.begin(), ivec2.end()));
    int largest = *(max_element(ivec2.begin(), ivec2.end()));

    cout << "Minsta talet i vectorn är: " << smallest << endl;
    cout << "Största talet i vectorn är: " << largest << endl << endl;


    // Sök efter inmatat tal i vectorn och skriv ut antalet förekomster
    int tal = 0, counter = 0;

    cout << "Ange ett tal att söka efter: ";
    cin >> tal;

    // Sök och räkna träffar, skriv ut resultat
    counter = count(ivec2.begin(), ivec2.end(), tal);
    cout << "Talet " << tal << " fanns " << counter  << " gånger i ivec2." << endl << endl;


    // Beräkna vilket nummer som förekommer flest gånger, samt hur många gånger
    int max = 0;
    int maxindex = ivec2[0];
    int numcount = 0;

    // Kan göras effektivare med en map, men det har vi inte gått igenom än. Just nu är det O(n²) men
    // skulle gå att modifera till O(n)
    for(auto i: ivec2)
    {
        numcount = (int)count(ivec2.begin(), ivec2.end(), ivec2[i]);
        if(numcount > max)
        {
            max = numcount;
            maxindex = ivec2[i];
        }
    }
    // Skriv ut resultat
    cout << "Talet med flest förekomster är: " << maxindex << " och förekommer " << max << " gånger." << endl << endl;


    // Sortera vectorn med <algorithm>
    sort(ivec2.begin(), ivec2.end());


    // Skriv ut alla ettor på en rad, tvåor på en rad osv
    // Metod 1 känns inte riktigt rätt, så jag skickar in två olika kodsnuttar
    cout << "Skriver ut vectorn med alla siffror på separata rader: " << endl;

    // Metod 1
    // Ny vektor för att lagra förekomsten/frekvensen
    //vector<int> frek(FREK, 0);
    // Fyll vektorn med nummerfrekvenserna
    //for(auto i = 0; i < MAX; i++)
    //    frek[ivec2[i]]++;

    // Skriv ut vektorns data rad för rad
    //for(i = 1; i < FREK + 1; i++)
    //{
    //    cout << i << " ";
    //    for(j = 0; j < frek[i]; j++)
    //        cout << i << " ";
    //    cout << endl;
    //}

    // Metod 2
    for(i = 0; i < MAX; i++)
    {
        int temp = ivec2[i];

        cout << ivec2[i] << " ";

        if(ivec2[i + 1] != temp)
            cout << endl;
    }


    return 0;
}


lab5_upg1_array/lab5_upg1_array.cpp
// lab5_upg1_array.cpp
// Program för att utföra ett antal operationer på en array
// Thomas Nordenmark 2013-12-12
// Ver 0.1
//------------------------------------------------------------------------------
#include <iostream> // cin, cout
#include <random> // default_random_engine
#include <iomanip> // setw, setprecision
#include <functional> // bind
using namespace std;

// =========================
// Deklaration av konstanter
// =========================
const int MAX=600; // Max platser i arrayen
const int FREK=100; // Antal platser i frekvensarray samt slumpgeneratorns max

// ======================
// Funktionsdeklarationer
// ======================
void bubbleSort(int arr[]); // Bubble sort

// ============
// Huvudprogram
// ============
int main()
{
    // Initiera slumpdistributörsobjektet generator
    default_random_engine generator(static_cast<unsigned>(time(0)));
    uniform_int_distribution<int> random(1, FREK); // Slumpa i itervallet 1-6
    // Slå ihop slumpgeneratorn till en funktion för enklare användning
    auto slumpaTal = bind(random, generator);

    int arr[MAX]; // Array med MAX antal platser
    // Iteratorer för loopar
    int i;
    //int j; // Behövs för utskriftmetod 1

    // Fyll arr[] med 600 slumpade tal mellan 1 och 100
    cout << "Slumpar 600 tal mellan 1 och 100 till arr[]." << endl << endl;
    for(i = 0; i < MAX; i++)
        arr[i] = slumpaTal(FREK);


    // Kopiera arr[] till arrcopy[]
    int arrcopy[MAX]; // Kopia av arr
    cout << "Kopierar arr[] till arrcopy[]." << endl << endl;
    for(i = 0; i < MAX; i++)
        arrcopy[i] = arr[i];


    // Beräkna medelvärdet av alla tal i arrayen
    int sum = 0; // Lagra arrayens totalsumma
    // Addera tal till totalsumman varje varv
    for(i = 0; i < MAX; i++)
        sum += arrcopy[i];

    double medel; // Lagra arrayens medelvärde
    medel = (double)sum / (double)MAX;  // Typecasta till double för högre precision
    cout << "Arrayens medelvärde är: " << fixed << setprecision(4) << medel << endl << endl; // Skriv ut medel med 2 decimaler


    // Sök efter minsta och största talet i arrayen
    int smallest = arrcopy[0]; // Sätt smallest till arrayens första element som start
    int largest = 0; // Sätt largest till 0 som utgångsläge

    // Iterera igenom arrayen, om ett tal är mindre än smallest, uppdatera smallest med det nya talet
    for(i = 1; i < MAX; i++)
    {
        if(arrcopy[i] < smallest)
            smallest = arrcopy[i];
        if(arrcopy[i] > largest)
            largest = arrcopy[i];
    }

    // Skriv ut det minsta  och största talet
    cout << "Minsta talet i arrayen är: " << smallest << endl;
    cout << "Största talet i arrayen är: " << largest << endl << endl;


    // Nu framgår det inte i vilken ordning operationerna ska utföras, själv hade jag sorterat arrayen
    // från första början för att förenkla alla andra operationer. T.ex. att hitta minsta och största
    // talet i den kommenterade koden nedan, men det känns för enkelt
    //
    //cout << "Det minsta talet är: " << arrcopy[0] << endl;
    //cout << "Det största talet är: " << arrcopy[MAX - 1] << endl;


    // Sök efter inmatat tal i arrayen och skriv ut antalet förekomster
    int tal = 0, counter = 0;
    cout << "Angel ett tal att söka efter: ";
    cin >> tal;

    // Sök efter det inmatade talet
    for(i = 0; i < MAX; i++)
        if(arrcopy[i] == tal)
            counter++;

    // Skriv ut antal förekomster för det inmatade talet
    cout << "Talet " << tal << " fanns " << counter << " gånger i arrcopy[]" << endl << endl;


    // Beräkna talens frekvens, s.k. histogram
    cout << "Beräknar sifferfrekvensen i arrcopy[]." << endl;

    // Initialisera en array frek[] med 100 platser för att lagra sifferfrekvens
    int frek[FREK];

    // Nollställ arrayen frek[] för att slippa smuts och konstiga beräkningar
    for(i = 0; i < MAX; i++)
        frek[i] = 0;

    // Iterera igenom arrcopy[] och uppdatera indexet i frek[] för varje nummer
    for(i = 0; i < MAX; i++)
        frek[arrcopy[i]]++;

    // Beräkna vilket nummer som förekommer flest gånger, samt hur många gånger
    int max = frek[0]; // Räknare för antalet förekomster
    int maxindex = 0; // Index för vilken position som innehåller det mest förekommande talet

    // Iterera igenom frekvens-arrayen
    for(i = 0; i < FREK; i++)
    {
      if(max < frek[i])
      {
        max = frek[i];
        maxindex = i;
      }
    }

    // Skriv ut tecknet med flest förekomster samt antal förekomster
    cout << "Tecknet med flest förekomster är: " << maxindex << " och förekommer " << max << " gånger." << endl << endl;


    // Sortera arrayen stigande med funktionen bubbleSort(). Jag skulle gjort det långt tidigare, för
    // att förenkla alla andra operationer, men det känns som det renderar en del av övningens moment
    // allt för enkla.
    // Med utskriftsmetod 1 nedan behövs ingen sortering alls
    cout << "Sorterar arrayen med bubble sort." << endl << endl;
    bubbleSort(arrcopy);


    // Skriv ut alla ettor på en rad, tvåor på en rad osv
    cout << "Skriver ut arrayen med alla siffror på separata rader: " << endl;

    // Metod 1 med extra array och dubbel for-loop
    //for(i = 1; i < FREK + 1; i++)
    //{
    //    cout << i << " ";
    //    for(j = 1; j < frek[i]; j++)
    //        cout << i << " ";
    //    cout << endl;
    //}

    // Metod 2 utan extra array
    for(i = 0; i < MAX; i++)
    {
        int temp = arrcopy[i];

        cout << arrcopy[i] << " ";

        if(arrcopy[i + 1] != temp)
            cout << endl;
    }

    return 0;
}

// =====================
// Funktionsdefinitioner
// =====================

// Bubble sort
void bubbleSort(int arr[])
{
    int i, j; // Iteratorer för looparna
    int temp; // Temporär lagringsplats för sortering

    // Yttre loop
    for(i = 0; i < MAX; i++)
    {
        // Inre loop
        for(j = 0; j < MAX - 1; j++)
        {
            if(arr[j] > arr[j + 1])
            {
                temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}
